### PR TITLE
SI-6891 Fix value class + tailrec crasher.

### PR DESCRIPTION
**Pull request by retronym**
rhs.substituteSymbols(old, new) leaves us with:

```
def loop#12225(x#12226: A#15491): scala#21.this.Unit#1615 =
  loop#12225(x#12226)
```

In which the TermSymbol x#12226 has a stale info, pointing at
the A#7274, the class type parameter, rather than A#15491,
the corresponding type parameter of the synthetic backing
method.

I've improved `TreeSymSubstituter` to substitute not
only `Tree#{tpe, symbol}`, but also `DefTree#sym.info`.

The `pos` test that triggered the new code path are
listed here: https://gist.github.com/4575687

Scathing review, please, by @paulp + @adriaanm; it's a small patch, but touches some fundamental machinery.

Obviously I could limit the scope of the new code to the `substituteSymbols` calls in `ExtensionMethods`, but I don't see how its correctness is context sensitive.

It seems to me that `substituteSymbols` was a breeding ground for inconsistencies between `tree.tpe` and `tree.symbol.info`, and I'm at a loss as to why that doesn't cause more problems.

I'm currently trying to add a new check to `TreeCheckers` to find references to type parameters and method parameters (to start with, but what else?) that don't respect scoping, ie their owner chain doesn't include the owner of the parameter. This check would have diagnosed this bug other like it. Any  and all suggestions for this line of investigation are most welcome. (Such checks will also be useful, nay, vital, for folks doing serious macro development)

### CODE REVIEW

**Comment by retronym** [comment]
An early version of said check:

https://github.com/retronym/scala/compare/scala:2.10.x...retronym:ticket/6891-checker

```
ticket/6891 ~/code/scala2 ./build/quick/bin/scalac -uniqid -Ycheck:extmethods test/files/pos/t6891.scala
[Now checking: extmethods]
[check: extmethods] The info of tree @scala#21.annotation#1622.tailrec#2397 def loop#12389(x#12390: A#15658): Unit#1613 = loop#12389(x#12390) refers to a out-of-scope type parameter type A#7437
[check: extmethods] The info of tree val x#12390: A#7437 = _ refers to a out-of-scope type parameter type A#7437
[check: extmethods] The info of tree loop#12389(x#12390) refers to a out-of-scope type parameter type A#7437
[check: extmethods] The info of tree loop#12389 refers to a out-of-scope type parameter type A#7437
[check: extmethods] The info of tree x#12390 refers to a out-of-scope type parameter type A#7437
[check: extmethods] The info of tree <synthetic> val x2#15642: O#7027.Foo#7436[A#7437] = (x1#15641.asInstanceOf#4726[O#7027.Foo#7436[A#15676]]: O#7027.Foo#7436[A#15676]) refers to a out-of-scope type parameter type A#7437
[check: extmethods] The info of tree <synthetic> val Foo$1#12663: O#7027.Foo#7436[A#7437] = x$1#15678.asInstanceOf#4726[O#7027.Foo#7436[A#15676]] refers to a out-of-scope type parameter type A#7437
[check: extmethods] The info of tree Foo$1#12663 refers to a out-of-scope type parameter type A#7437
```

(Edit: that output is from the test case for this bug, without the patch to substitution applied.)

**Comment by paulp** [comment]
It's so nice seeing someone else running across stuff like this and saying "I'm at a loss as to why that doesn't cause more problems."

File: src/reflect/scala/reflect/internal/Trees.scala [hunk 1]
```diff
@@ -1440,6 +1440,18 @@ trait Trees extends api.Trees { self: SymbolTable =>
       if (tree.hasSymbol) {
         subst(from, to)
         tree match {
+          case _: DefTree =>
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Trees.scala:1443]: Is this sufficient in light of these revelations? https://groups.google.com/forum/?fromgroups#!topic/scala-internals/WXpouzb_g_E

**Comment by retronym** [comment]
I just ran this patch by @odersky, he agreed that it appears to correct an oversight. He suggests that we should also treat `Function` and `Return` nodes, as in the `ChangeOwnerTraverser`. I'll try to craft test cases to motivate those.

We also discussed the `TreeChecker` enhancement: in addition to checking that type parameters and method parameters are only referenced from within their owning entity, we can also check that any symbol owned by a term should satisfy the same property. This would catch local definitions. I'll roll those checks into another branch, and run a build with them turned on to shine a light on non-compliant trees.

File: src/reflect/scala/reflect/internal/Trees.scala (hunk 1 above)
Comment (paulp) [inline src/reflect/scala/reflect/internal/Trees.scala:1443]: Well, if there IS a relevant quality captured by "is a DefTree, Return, Function, or Import" then it's about time to encapsulate it. Do you want to take this on a spin with that broader view? I'd be curious what falls in there and whether we can use that info to construct failures.

Comment (retronym) [inline src/reflect/scala/reflect/internal/Trees.scala:1443]: Will try. So far I've found `ChangeOwnerTraverser`.

Comment (retronym) [inline src/reflect/scala/reflect/internal/Trees.scala:1443]: As best as I can figure: `Function` and `Import` define symbols, but they don't have interesting `info`s. `Return` reference enclosing method symbols, and are treated by `ChangeOwnerTraverser`. None of them appears to be relevant to this chunk of code.

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by adriaanm** [comment]
LGTM'ed by odersky

**Pull request merged**